usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -9585,3 +9585,110 @@ static void free_ep11_target_for_apqn(target_t target)
     }
 }
 
+CK_RV token_specific_set_attribute_values(STDLL_TokData_t *tokdata, OBJECT *obj,
+                                          TEMPLATE *new_tmpl)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_OBJECT_CLASS class;
+    size_t keyblobsize = 0;
+    CK_BYTE *keyblob;
+    DL_NODE *node;
+    CK_ATTRIBUTE *ibm_opaque_attr = NULL;
+    CK_ATTRIBUTE_PTR attributes = NULL;
+    CK_ULONG num_attributes = 0;
+    CK_ATTRIBUTE *attr;
+    CK_RV rc;
+
+    rc = template_attribute_get_ulong(obj->template, CKA_CLASS, &class);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s CKA_CLASS is missing\n", __func__);
+        return rc;
+    }
+
+    switch (class) {
+    case CKO_SECRET_KEY:
+    case CKO_PRIVATE_KEY:
+        break;
+    default:
+        /* Not a secret or private key, nothing to do */
+        return CKR_OK;
+    }
+
+    rc = obj_opaque_2_blob(tokdata, obj, &keyblob, &keyblobsize);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s no key-blob rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    node = new_tmpl->attribute_list;
+    while (node) {
+        attr = (CK_ATTRIBUTE *)node->data;
+
+        /* EP11 can set certain boolean attributes only */
+        switch (attr->type) {
+        case CKA_EXTRACTABLE:
+        case CKA_MODIFIABLE:
+        case CKA_SIGN:
+        case CKA_SIGN_RECOVER:
+        case CKA_DECRYPT:
+        case CKA_ENCRYPT:
+        case CKA_DERIVE:
+        case CKA_UNWRAP:
+        case CKA_WRAP:
+        case CKA_VERIFY:
+        case CKA_VERIFY_RECOVER:
+        case CKA_WRAP_WITH_TRUSTED:
+        case CKA_TRUSTED:
+        case CKA_IBM_RESTRICTABLE:
+        case CKA_IBM_USE_AS_DATA:
```

Comment:
There are additional boolean attributes, which can be changed, e.g. CKA_COPYABLE and CKA_DESTROYABLE. Did you intentionally leave them out?

Reply:
I took those that are listed in the EP11 structure document under 2.2.6 Boolean attribute listing. And CKA_COPYABLE and CKA_DESTROYABLE are not listed there. 
Not sure if the doc just does not mention them, but would support them anyway, or if they really don't support them.